docs/docs/forms.mdx

```diff
@@ -360,74 +360,99 @@ to fully customize the validation process.
-which slots should be requested. Rasa Open Source will run your custom action whenever
-the form validates user input. Set the slot `requested_slot` to the name of the slot
-which should be extracted next. If all desired slots are filled, set `requested_slot`
-to `None`.
+### Dynamic Form Behavior
 
-The following example shows the implementation of a custom action which requests
-the three slots `cuisine`, `num_people`, and `outdoor_seating`.
+By default Rasa Open Source will ask for the next
+[`requested_slot`](forms.mdx#the-requested_slot-slot) defined in the slot mappings
+in the domain, which is not filled yet. If you use
+[custom slot mappings](forms.mdx#custom-slot-mappings) and the `FormValidationAction`,
+it will ask for the first empty slot returned by the `required_slots` method. If all
+slots in `required_slots` are filled the form will be be deactivated.
 
-```python
-from typing import Dict, Text, List
+If needed, you can update the required slots of your form dynamically.
+This is, for example, useful when you need further details based on how
+a previous slot was filled or you want to change the order in which slots are requested.
 
-from rasa_sdk import Tracker
-from rasa_sdk.events import EventType
-from rasa_sdk.executor import CollectingDispatcher
-from rasa_sdk import Action
-from rasa_sdk.events import SlotSet
+If you are using the Rasa SDK, we recommend you to use the `FormValidationAction` and
+override `required_slots` to fit your dynamic behavior.
+The example below will ask the user if they want to sit in
+the shade or in the sun in case they told us that they want to sit outside.
 
+```python
+from typing import Text, List, Optional
+
+from rasa_sdk.forms import FormValidationAction
 
-class ValidateRestaurantForm(Action):
+class ValidateRestaurantForm(FormValidationAction):
     def name(self) -> Text:
         return "validate_restaurant_form"
 
-    def run(
-        self, dispatcher: CollectingDispatcher, tracker: Tracker, domain: Dict
-    ) -> List[EventType]:
-        required_slots = ["cuisine", "num_people", "outdoor_seating"]
-
-        for slot_name in required_slots:
-            if tracker.slots.get(slot_name) is None:
-                # The slot is not filled yet. Request the user to fill this slot next.
-                return [SlotSet("requested_slot", slot_name)]
-
-        # All slots are filled.
-        return [SlotSet("requested_slot", None)]
+    async def required_slots(
+        self,
+        slots_mapped_in_domain: List[Text],
+        dispatcher: "CollectingDispatcher",
+        tracker: "Tracker",
+        domain: "DomainDict",
+    ) -> Optional[List[Text]]:
+        additional_slots = ["outdoor_seating"]
+        if tracker.slots.get("outdoor_seating") is True:
+            # If the user wants to sit outside, ask
+            # if they want to sit in the shade or in the sun.
+            additional_slots.append("shade_or_sun")
+
+        return additional_slots + slots_mapped_in_domain
```

Comment:
I'm a huge fan of this <3

Reply:
Though we should also add a note or a warning here, reminding the user to implement `extract_<slot_name>` when a new slot is added that isn't mapped in the domain.

Reply:
I mean in the docs, not a warning in the code (there, it would be an error :-D).